File: c/problem_18.c
#include "problem_18.h"
#include <stdio.h>
#include <assert.h>

#define MAX_N 256

/* maxSubarrayK: monotonic deque approach. O(n) time, O(k) space. */
static void maxSubarrayK(const int *nums, int n, int k, int *out, int *outLen) {
    int deque[MAX_N]; /* stores indices */
    int front = 0, back = 0; /* [front, back) */
    *outLen = 0;

    for (int i = 0; i < n; i++) {
        while (front < back && deque[front] < i - k + 1) front++;
        while (front < back && nums[deque[back - 1]] < nums[i]) back--;
        deque[back++] = i;
        if (i >= k - 1) out[(*outLen)++] = nums[deque[front]];
    }
}

void problem18(void) {
    printf("Problem 18:\n");
    int nums[] = {10,5,2,7,8,7};
    int out[MAX_N], outLen;
    int exp[] = {10,7,8,8};
    maxSubarrayK(nums, 6, 3, out, &outLen);
    assert(outLen == 4);
    for (int i = 0; i < 4; i++) assert(out[i] == exp[i]);
    printf("All Tests Passed\n");
}

File: rust/problem_18.rs
use std::collections::VecDeque;

// max_subarray_of_size_k returns the maximum value in each sliding window of size k.
// Uses a monotonic deque of indices in decreasing value order.
// This runs in O(n) time and O(k) space.
pub fn max_subarray_of_size_k(nums: &[i64], k: usize) -> Vec<i64> {
    let mut deque: VecDeque<usize> = VecDeque::new();
    let mut result = Vec::new();

    for i in 0..nums.len() {
        // remove indices outside the window
        while deque.front().map_or(false, |&f| f + k <= i) {
            deque.pop_front();
        }
        // remove smaller elements from the back
        while deque.back().map_or(false, |&b| nums[b] < nums[i]) {
            deque.pop_back();
        }
        deque.push_back(i);
        if i + 1 >= k { result.push(nums[*deque.front().unwrap()]); }
    }
    result
}

pub fn problem18() {
    println!("Problem 18:");
    assert_eq!(max_subarray_of_size_k(&[10,5,2,7,8,7], 3), vec![10,7,8,8], "assertion failed");
    assert_eq!(max_subarray_of_size_k(&[1,2,3], 1), vec![1,2,3], "assertion failed");
    println!("All Tests Passed");
}
